pep-0706.rst

```diff
@@ -0,0 +1,509 @@
+- Unpacking a general data archive: the ``shutil.unpack_archive`` use case,
+  where it's not important to preserve details specific to ``tar`` or
+  Unix-like filesystems.
+
+After a deprecation period, the last option -- the most limited
+but most secure one -- will become the default.
+
+Even with better general defaults, users should still verify the archives
+they extract, and perhaps modify some of the metadata.
+Superficially, the following looks like a reasonable way to do this today:
+
+* Call ``TarFile.getmembers()``
+* Verify or modify each member's ``TarInfo``
+* Pass the result to ``extractall``'s ``members``
+
+However, there are some issues with this approach:
+
+- ``TarInfo`` does not officially allow modification. You can modify it, but you
+  can get surprising results since ``TarInfo`` objects are cached/shared.
+- Calling ``getmembers`` can be expensive and it 
+  `requires a seekable archive <https://github.com/python/cpython/issues/45385#issuecomment-1255615199>`__.
+- When verifying members in advance, it may be necessary to track how each
+  member would have changed the filesystem, e.g. how symlinks are being set up.
+  This is hard. We can't expect users to do it.
+
+To solve these issues we'll:
+
+- Provide a supported way to modify ``TarInfo`` when needed,
+  without unnecessary work when not.
+  A ``replace`` method, similar to `dataclass.replace <https://docs.python.org/3/library/dataclasses.html#dataclasses.replace>`_
+  or `namedtuple._replace <https://docs.python.org/3/library/collections.html#collections.somenamedtuple._replace>`_,
+  should do the trick.
+- Provide a “filter” hook in ``extractall``'s loop that can modify or discard
+  members before they are processed.
+- Require that this hook is called just before extracting each member,
+  so it can scan the *current* state of the disk. This will greatly simplify
+  the implementation of policies (both in stdlib and user code),
+  at the cost of not being able to do a precise “dry run”.
+
+The hook API will be very similar to the existing ``filter`` argument
+for ``TarFile.add``.
+We'll also name it ``filter``.
+(In some cases “policy” would be a more fitting name,
+but the API can be used for more than security policies.)
+
+The built-in policies/filters described above will be implemented using the
+public filter API, so they can be used as building blocks or examples.
+
+
+Setting a precedent
+-------------------
+
+If and when other libraries  archive extraction, such as ``zipfile``,
+gain similar functionality, they should mimic the API as much as it's
+reasonable.
+
+To enable this for simple cases, the built-in filters will have string names,
+e.g. users can pass ``filter='data'`` instead of a specific function that deals
+with ``TarInfo`` objects.
+
+The ``shutil.extract_archive()`` function will get a ``filter`` argument,
+which it will pass to ``extractall``.
+
+Adding *functional* API that would work across archive formats is
```

Comment:
```suggestion
Adding a *functional* API that would work across archive formats is
```
Also, I'm not clear what "functional" means here. If it's just "functioning", you may be better off dropping the word. If it's something to do with functions, I don't understand what it means :)

Reply:
One that takes functions rather than strings. I guess:
```suggestion
Adding function-based API that would work across archive formats is
```